### ISSUE TITLE
Reconstructing docs from query indices

### ISSUE THREAD

**Issue by jarohen** [xtdb/xtdb#1511]
Currently, when required to reconstruct documents (e.g. for `entity` or `pull [*]`) we have to go to the document store, because the query indices don't contain enough information. Specifically, they cannot reconstruct values that are sets or vectors - e.g. if we find a single value in the ECAV index we don't know whether it's actually a single value or a single element set/vector, and we don't store the ordering/duplicate values of vectors.

For remote/pay-per-req doc-stores, we do recommend as large a local doc-store cache as is practical, although this is mostly a duplication of data that's already stored in local query indices.

Proposal:
- the values in the ECAV index are unused - we could store information here about whether the value is a scalar, a set, or its indices within a vector.
- the attribute hashes in the ECAV index currently aren't reversible - we'll have to either make these reversible, or keep a cached mapping from the `all-attrs` function

**Comment by jarohen** [comment]
Part one of this (the changes to the indices) was merged in f5ee85bf15abc4843f93b88bae7e33d0ab146ecd - this means that the change to enable it shouldn't require an index version bump (or even, indeed, a breaking API change). Unfortunately enabling it wasn't adequately performant, so I've parked that for now in the interest of unblocking the 1.17 release.

See #1512 for more details.

**Comment by refset** [comment]
Noting that in the event of a ~catastrophic document store failure that these merged Part One changes could be useful for some forensic investigation/recovery assuming valid index-stores/checkpoints exist.

**Comment by refset** [comment]
Noting from a conversation with @jarohen last week that there is an aspect of the current indexed representation which would need resolving as part of this issue: the case where an original vector value contains duplicate entries (IIRC).
